db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,123 @@
+    , _snapshot_name(std::move(snapshot_name))
+{}
+
+std::string backup_task_impl::type() const {
+    return "backup";
+}
+
+tasks::is_internal backup_task_impl::is_internal() const noexcept {
+    return tasks::is_internal::no;
+}
+
+future<> backup_task_impl::run(sstring data_dir) {
+    return async([this, data_dir = std::move(data_dir)] {
+        gate uploads;
+        auto wait = defer([&uploads] { uploads.close().get(); });
+        std::exception_ptr ex;
+        //
+        // The keyspace data directories and their snapshots are arranged as follows:
+        //
+        //  <data dir>
+        //  |- <keyspace name1>
+        //  |  |- <column family name1>
+        //  |     |- snapshots
+        //  |        |- <snapshot name1>
+        //  |          |- <snapshot file1>
+        //  |          |- <snapshot file2>
+        //  |          |- ...
+        //  |        |- <snapshot name2>
+        //  |        |- ...
+        //  |  |- <column family name2>
+        //  |  |- ...
+        //  |- <keyspace name2>
+        //  |- ...
+        //
+        auto ks_dir = fs::path(data_dir) / _ks;
+        auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+        auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+        snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+        while (auto table_ent = ks_dir_lister.get().get()) {
+            auto snapshot_dir = ks_dir / table_ent->name / sstables::snapshots_dir / _snapshot_name;
+            auto has_snapshot = file_exists(snapshot_dir.native()).get();
+
+            snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
+            if (!has_snapshot) {
+                continue;
+            }
+
+            auto cf_name_and_uuid = replica::parse_table_directory_name(table_ent->name);
+            auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+            auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
+
+            while (auto component_ent = snapshot_dir_lister.get().get()) {
+                auto gh = uploads.hold();
+                auto component_name = snapshot_dir / component_ent->name;
+                auto destination = fmt::format("/{}/{}/{}/{}", _bucket, cf_name_and_uuid.first, _snapshot_name, component_ent->name);
+                snap_log.trace("Upload {} to {}", component_name.native(), destination);
+                // Start uploading in the background. The caller waits for these fibers
+                // with the _uploads gate.
+                // Parallelizm is implicitly controlled in two ways:
+                //  - s3::client::claim_memory semaphore
+                //  - http::client::max_connections limitation
+                std::ignore = _client->upload_file(component_name, destination).handle_exception([this, comp = component_name] (auto ex) {
```

Comment:
Noob question, is there retries in case of first exception thrown during a call to `upload_file`?

Reply:
Not yet :(